src/firmware/boards/powerboard_esp32/PowerLoop.cpp

```diff
@@ -0,0 +1,275 @@
+#include <SoftwareSerial.h> // include library to enable software serial ports
+#include <EEPROM.h> // include library to read and write from flash memory
+
+#define EEPROM_SIZE 4
+
+SoftwareSerial SUART(2,3); //SRX = DPin-2; STX = DPin-3
+
+//Message Storage
+char INITIAL_MESSAGE;
+char MESSAGE;
+
+//Action Messages Library
+char HV_MEASUREMENT;
+char GENEVA_ENCODER;
+char GENEVA_MOTOR_ANGLE;
+char GENEVA_MOTOR_SPEED;
+char BREAKBEAM;
+char CHICKER_AUTO_KICK;
+char CHICKER_AUTO_CHIP;
+char CHICKER_KICK_DISTANCE_SPEED;
+char CHICKER_CHIP_DISTANCE_SPEED;
+char PULSE_WIDTH;
+char FLYBACK;
+
+//PINS
+const int HV_MEASUREMENT_PIN;
+const int G_ENCODER_PIN;
+const int G_MOTOR_ANGLE_PIN;
+const int G_MOTOR_SPEED_PIN;
+const int BREAKBEAM_INTER_PIN;
+const int BREAKBEAM_DIFF_PIN;
+const int CHICKER_AUTOKICK_PIN;
+const int CHICKER_AUTOCHIP_PIN;
+const int KICK_DISTANCE_SPEED_PIN;
+const int KICK_DISTANCE_SPEED_VAL_PIN;
+const int CHIP_DISTANCE_SPEED_PIN;
+const int CHIP_DISTANCE_SPEED_VAL_PIN;
+const int PULSE_WIDTH_PIN;
+const int FLYBACK_FAULT_PIN;
+
+//Flash Address
+int HV_MEASUREMENT_ADDRESS;
+int G_ENCODER_ADDRESS;
+int G_MOTOR_ANGLE_ADDRESS;
+int G_MOTOR_SPEED_ADDRESS;
+int BREAKBEAM_INTER_ADDRESS;
+int BREAKBEAM_DIFF_ADDRESS;
+int CHICKER_AUTOKICK_ADDRESS;
+int CHICKER_AUTOCHIP_ADDRESS;
+int KICK_DISTANCE_SPEED_ADDRESS;
+int CHIP_DISTANCE_SPEED_ADDRESS;
+int PULSE_WIDTH_ADDRESS;
+int FLYBACK_FAULT_ADDRESS;
+
+//Others
+int COUNTER;
+float checker;
+
+int hv_data;
+int g_encoder_angle;
+float g_motor_angle;
+float g_motor_speed;
+int breakbeam_diff;
+int breakbeam_inter;
+int autokick;
+int autochip;
+float kick_distance_speed;
+float chip_distance_speed;
+int flyback_fault;
+
+
+
+void setup(){
+
+    //Initializing serial ports 
+    Serial.begin(9600);
+    SUART.begin(9600);
+
+    //Initializing GPIO pins
+    pinMode(HV_MEASUREMENT_PIN, INPUT);
+    pinMode(G_ENCODER_PIN, INPUT);
+    pinMode(G_MOTOR_ANGLE_PIN, INPUT);
+    pinMode(G_MOTOR_SPEED_PIN, INPUT);
+    pinMode(BREAKBEAM_DIFF_PIN, INPUT);
+    pinMode(BREAKBEAM_INTER_PIN, INPUT);
+    pinMode(CHICKER_AUTOCHIP_PIN, INPUT);
+    pinMode(CHICKER_AUTOKICK_PIN, INPUT);
+    pinMode(KICK_DISTANCE_SPEED_PIN, INPUT);
+    pinMode(CHIP_DISTANCE_SPEED_PIN, INPUT);
+    pinMode(PULSE_WIDTH_PIN, INPUT);

```

Comment:
you need 2 of these, 1 for kicker solenoid and 1 for chipper solenoid, and they should be outputs